### PR TITLE
fix: remove layer_removal_cs

### PR DESCRIPTION
**Pull request by koivunej**
Quest: https://github.com/neondatabase/neon/issues/4745. Follow-up to #4938.

- add in locks for compaction and gc, so we don't have multiple executions at the same time in tests
- remove layer_removal_cs
- remove waiting for uploads in eviction/gc/compaction
    - #4938 will keep the file resident until upload completes

### CODE REVIEW

File: pageserver/src/tenant/timeline.rs [hunk 1]
```diff
@@ -1111,16 +1110,6 @@ impl Timeline {
...
     ) -> anyhow::Result<Vec<Option<Result<(), EvictionError>>>> {
-        // ensure that the layers have finished uploading
-        // (don't hold the layer_removal_cs while we do it, we're not removing anything yet)
-        remote_client
-            .wait_completion()
-            .await
-            .context("wait for layer upload ops to complete")?;
-
-        // now lock out layer removal (compaction, gc, timeline deletion)
-        let _layer_removal_guard = self.layer_removal_cs.lock().await;
```
Comment (jcsp) [inline pageserver/src/tenant/timeline.rs:1122]: So I guess we would now block on `evict_and_wait` if a compaction was going on and touched some of the layers we're about to evict, right?

This may be a bit OT for this PR, but it makes me wonder if we should make the wanted_evicted field hold a timestamp (the access time when eviction decided to evict the layer), and if the layer has been accessed more recently then we shouldn't go through with  the eviction.  That's obviously an optimization thing rather than a safety thing though.

Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:1122]: > This may be a bit OT for this PR, but it makes me wonder if we should make the wanted_evicted field hold a timestamp (the access time when eviction decided to evict the layer), and if the layer has been accessed more recently then we shouldn't go through with the eviction. That's obviously an optimization thing rather than a safety thing though.

I did consider that but it is difficult to say would this help. This would move the decision to evict or not to Layer, which I am not keen on doing because you cannot have a full picture there. Theoretically the value is already there. I could add a warning metric for when eviction happens to within one minute accessed Layer.

Currently evictions always lose to re-downloads, meaning that we could have:
1. Layer::evict_and_wait
2. last `Arc<DownloadedLayer>` is dropped, `spawn_blocking` is scheduled
3. before `spawn_blocking` gets to run, there's a new read access to layer contents which gets to run first incrementing the version -- it finds the file locally
4. when spawn_blocking finally gets to run, we notice the version mismatch and eviction is cancelled.

Could be the step 4 is actually a bit racy now that I think of it. Need to check that as well. However this is as designed and as I would like to keep it. Perhaps I just need to get rid of the version field and the "rare metric".

Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:1122]: To the original:
> So I guess we would now block on evict_and_wait if a compaction was going on and touched some of the layers we're about to evict, right?

No blocking needed. We would just evict or fail to evict regardless of future compaction access (which would start by first acquiring RAII guards keeping the layers resident -- `ResidentLayer`).

Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:1122]: Found a definitive forever busy loop possibility in case the spawn_blocking never runs, and before that it will be a busy loop.

Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:1122]: Coming back to this after #5742, #5743, #5744, #5756, #5766, which were pending anyways.

**Comment by koivunej** [comment]
After fixing the RemoteTimelineClient at least hastily, the scrubber seems to cause errors for any unlinked file?

File: pageserver/src/deletion_queue/validator.rs [hunk 2]
```diff
@@ -179,6 +179,11 @@ where
 
             if valid && *validated_generation == tenant_lsn_state.generation {
                 for (_timeline_id, pending_lsn) in tenant_lsn_state.timelines {
+                    tracing::info!(
+                        "{_timeline_id}: {} => {}",
+                        pending_lsn.result_slot.load(),
+                        pending_lsn.projected
+                    );
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/deletion_queue/validator.rs:186]:
```suggestion
                    tracing::debug!(
                        "{_timeline_id}: {} => {}",
                        pending_lsn.result_slot.load(),
                        pending_lsn.projected
                    );
```

File: test_runner/regress/test_remote_storage.py [hunk 3]
```diff
@@ -784,91 +782,86 @@ def test_compaction_waits_for_upload(
...
-    except Exception:
-        # Log the actual failure's backtrace here, before we proceed to join threads
-        log.exception("Failure, cleaning up...")
-        raise
-    finally:
-        compact_barrier.abort()
+    compacted_layers = client.layer_map_info(tenant_id, timeline_id).historic_by_name()
+
+    overlap = compacted_layers.intersection(upload_stuck_layers)
+    assert len(overlap) == 0, "none of the L0's should remain after L0 => L1 compaction"
```
Comment (problame) [inline test_runner/regress/test_remote_storage.py:836]: Should also assert that there are some *new* *L1* layers in the `compacted_layers`.
As a self-check to ensure that we've indeed done a L0=>L1 compaction.

File: test_runner/regress/test_remote_storage.py [hunk 4]
```diff
@@ -784,91 +782,86 @@ def test_compaction_waits_for_upload(
...
 
     # Ensure that this actually terminates
     wait_upload_queue_empty(client, tenant_id, timeline_id)
 
+    def until_gcs_completed():
+        gcs = gcs_completed()
+        log.info(f"gcs: {gcs}")
+        assert gcs >= len(upload_stuck_layers)
+
+    wait_until(10, 1, until_gcs_completed)
```
Comment (problame) [inline test_runner/regress/test_remote_storage.py:857]: I don't see us do any garbage collection here, which I assume `gcs` means (?)

File: s3_scrubber/src/checks.rs [hunk 5]
```diff
@@ -136,24 +135,17 @@ pub(crate) async fn branch_cleanup_and_check_errors(
...
-                            // be something enqueued for deletion, so while this check is valid 
+                            // be something enqueued for deletion, so while this check is valid
                             // for indicating that a layer is garbage, it is not an indicator
                             // of a problem.
                             gen < &index_part_generation)
                         .collect();
 
+                    // unlinked files at the current generation are not considered to be deleted,
+                    // because a pageserver could still have those in its deletion queue or even as
+                    // layers which are being read.
```
[the remark this answers is not in the document]
Comment (koivunej) [reply s3_scrubber/src/checks.rs:146]: > IMO if the layer file is removed from the IndexPart, it's long gone from disk.

This is incorrect. When gc and compaction complete, they now only unlink the file from index_part.json as the file is logically gone, but we might still have `Layer` values for those files. We could still have the file on-disk, or we might still on-demand download the file. This follows from "you have a Layer, you can read it".

Unlinking and explicit "danglingness" was introduced in #5645, compaction support in #5649 and #5675.

File: pageserver/src/tenant/mgr.rs [hunk 6]
```diff
@@ -1844,13 +1845,42 @@ pub(crate) async fn immediate_gc(
...
+                    }
+                }
+
+                let timeline = tenant.get_timeline(timeline_id, false).ok();
+                let rtc = timeline.as_ref().and_then(|x| x.remote_client.as_ref());
+
+                if let Some(rtc) = rtc {
+                    // layer drops schedule actions on remote timeline client to actually do the
+                    // deletions; don't care just exit fast about the shutdown error
+                    drop(rtc.wait_completion().await);
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs:1880]: I think it has happened due to coevolution of deletionqueue tests. There are tests which assume that after `/do_gc` all deletions have been scheduled and sent over to deletionqueue so the test can call `flush` on the deletionqueue.

Removing the layer_removal_cs makes it more likely that there are still some references to the layers OR within test suite execution the spawn_blocking gets unlucky long execution times.

File: pageserver/src/tenant/remote_timeline_client.rs [hunk 7]
```diff
@@ -841,6 +840,63 @@ impl RemoteTimelineClient {
...
+
+    fn schedule_shutdown(
+        self: &Arc<Self>,
+        upload_queue: &mut UploadQueueInitialized,
+    ) -> tokio::sync::oneshot::Receiver<()> {
+        let (tx, rx) = tokio::sync::oneshot::channel();
+        let op = UploadOp::Shutdown(Some(tx));
+
+        upload_queue.queued_operations.push_back(op);
+        // this operation is not counted either
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client.rs:893]: I was thinking of the barrier op but will adjust the comment.

File: pageserver/src/tenant/remote_timeline_client.rs [hunk 8]
```diff
@@ -1084,6 +1142,19 @@ impl RemoteTimelineClient {
                 break;
             }
 
+            if let UploadOp::Shutdown(maybe_tx) = next_op {
+                if let Some(tx) = maybe_tx.take() {
+                    if tx.send(()).is_err() {
+                        tracing::error!(
+                                "RemoteTimelineClient::shutdown has been cancelled; this should never happen."
+                        );
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client.rs:1150]: Usually rustfmt likes to split, but here there is def extra indentation.

File: pageserver/src/tenant/timeline.rs [hunk 9]
```diff
@@ -3954,6 +3896,9 @@ impl Timeline {
             //
             // This does not in fact have any effect as we no longer consider local metadata unless
             // running without remote storage.
+            //
+            // This unconditionally schedules also an index_part.json update, even though, we will
+            // be doing one a bit later with the unlinked gc'd layers.
             self.update_metadata_file(self.disk_consistent_lsn.load(), None)
                 .await?;
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3903]: I think this was required by why test_gc_cutoff was introduced in #2558 which was introduced to fix #2539. Also I think #5624 should just be closed as duplicate.

Reading Kirills comment and this comment of yours I think this can now be removed: https://github.com/neondatabase/neon/issues/2539#issuecomment-1265801283

File: test_runner/regress/test_remote_storage.py [hunk 10]
```diff
@@ -784,91 +782,86 @@ def test_compaction_waits_for_upload(
...
         wait_for_last_flush_lsn(env, endpoint, tenant_id, timeline_id)
 
-        # Now make the flushing hang and update one small piece of data
-        client.configure_failpoints(("before-upload-layer-pausable", "pause"))
+        if deltas_at_first < 2:
+            # why complicate with conditionals? we might fix the initdb image layer optimization
+            client.timeline_checkpoint(tenant_id, timeline_id)
+            deltas_at_second = len(client.layer_map_info(tenant_id, timeline_id).delta_layers())
+            assert deltas_at_second > deltas_at_first
+            assert deltas_at_second == 2
```
[the remark this answers is not in the document]
Comment (koivunej) [reply test_runner/regress/test_remote_storage.py:808]: Created #5863 some time ago, will link here.

File: test_runner/regress/test_remote_storage.py [hunk 11]
```diff
@@ -784,91 +782,86 @@ def test_compaction_waits_for_upload(
...
-            log.info("Compaction errored: {e}")
-            compact_result.put(e)
+    for name in upload_stuck_layers:
+        path = env.pageserver.timeline_dir(tenant_id, timeline_id) / name
+        assert path.exists(), "while uploads are stuck the layers should be present on disk"
 
-    checkpoint_thread = threading.Thread(target=checkpoint_in_background)
-    checkpoint_thread.start()
+    # now this will do the L0 => L1 compaction and want to remove our layers
+    client.timeline_checkpoint(tenant_id, timeline_id)
```
[the remark this answers is not in the document]
Comment (koivunej) [reply test_runner/regress/test_remote_storage.py:826]: Yep, as my PR to delete the optimization path was not included #5843 (probably resulted in me creating the issue #5863) the complication can now be removed and just have it break via asserting that we do in fact have enough layers now.

File: test_runner/regress/test_remote_storage.py (hunk 3 above)
Comment (koivunej) [reply test_runner/regress/test_remote_storage.py:836]: I think we can assert that it's one L1 here, makes sense.

File: pageserver/src/tenant/timeline/delete.rs [hunk 12]
```diff
@@ -156,9 +156,9 @@ async fn delete_local_layer_files(
     tenant_id: TenantId,
     timeline: &Timeline,
 ) -> anyhow::Result<()> {
-    info!("waiting for layer_removal_cs.lock()");
-    let layer_removal_guard = timeline.layer_removal_cs.lock().await;
-    info!("got layer_removal_cs.lock(), deleting layer files");
+    info!("waiting for locks");
+    let guards = tokio::join!(timeline.gc_lock.lock(), timeline.compaction_lock.lock());
+    info!("got locks, deleting layer files");
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline/delete.rs:161]: > does tokio::join! poll the futures in any particular order

no order. I would expect it to lower to select, but did not check.

File: test_runner/regress/test_remote_storage.py (hunk 4 above)
Comment (problame) [reply test_runner/regress/test_remote_storage.py:857]: I remember objecting to calling the "dangling file deletion" "garbage collection" :/

File: pageserver/src/tenant/timeline.rs [hunk 13]
```diff
@@ -3478,34 +3461,16 @@ impl Timeline {
...
-        // See remote_timeline_client module level comment on consistency.
-        // Do it here because we don't want to hold self.layers.write() while waiting.
-        if let Some(remote_client) = &self.remote_client {
-            debug!("waiting for upload ops to complete");
-            remote_client
-                .wait_completion()
-                .await
-                .context("wait for layer upload ops to complete")?;
-        }
-
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3599]: This was discussed at the offsite and in general we are in agreement that this could and probably should be done. (creating an issue.)

[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3599]: #5897

File: pageserver/src/tenant/timeline.rs (hunk 1 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:1203]: This question was about wanted_evicted (enum variant and an atomic); it has now been implemented as eviction cancelling on access; we have non-zero occurances of those in production. Resolving.

File: pageserver/src/deletion_queue/validator.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/deletion_queue/validator.rs:186]: In "chore: cleanup added validator logging".

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client.rs:893]: "doc: make the comment better" or b05d8d17c later

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client.rs:1150]: "chore: fix formatting" or 22314b05f later.

File: pageserver/src/tenant/timeline/delete.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline/delete.rs:161]: These locks are introduced in this PR and they are guarding against http apis and background task executing the compaction and gc.

I'm adding comments about how these can only be used for these things; I don't think there is an abstraction which would allow guarding against the misuse...

Timeline deletion does stop all tasks before, so it is only guarding against the http api there as far as I can see.

6505dca8b doc: guidance against using per-op locks for other things

File: pageserver/src/tenant/mgr.rs [hunk 14]
```diff
@@ -1836,6 +1836,7 @@ pub(crate) async fn immediate_gc(
     // Run in task_mgr to avoid race with tenant_detach operation
     let ctx = ctx.detached_child(TaskKind::GarbageCollector, DownloadBehavior::Download);
     let (task_done, wait_task_done) = tokio::sync::oneshot::channel();
+    // TODO: spawning is redundant now, need to hold the gate
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs:1839]: I just learned `do_gc` is enabled in production builds, so I will need to clean the TODO's around here.

I think the spawning one is still valid, but I did not introduce it here.

File: pageserver/src/tenant/timeline.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3903]: But yeah, as we still don't require #4099 cannot remove this. (Even if the non-remote_storage one is currently broken.)

[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3903]: `doc: cleanup, link 4099` perhaps fbba48830

File: test_runner/regress/test_remote_storage.py (hunk 4 above)
Comment (koivunej) [reply test_runner/regress/test_remote_storage.py:857]: So the span would be "layer_(un|de)dangle"? :) Perhaps instead `layer_delete` as that is what we are calling it over at RemoteTimelineClient.

File: test_runner/regress/test_remote_storage.py [hunk 15]
```diff
@@ -784,91 +782,86 @@ def test_compaction_waits_for_upload(
     timeline_id = env.initial_timeline
 
     client = env.pageserver.http_client()
+    layers_at_creation = client.layer_map_info(tenant_id, timeline_id)
+    deltas_at_creation = len(layers_at_creation.delta_layers())
+
+    # Now make the flushing hang
+    client.configure_failpoints(("before-upload-layer-pausable", "pause"))
```
[the remark this answers is not in the document]
Comment (koivunej) [reply test_runner/regress/test_remote_storage.py:789]: I did note this later on, but let's move it upwards.

[the remark this answers is not in the document]
Comment (koivunej) [reply test_runner/regress/test_remote_storage.py:789]: b5ab83a14

File: test_runner/regress/test_remote_storage.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply test_runner/regress/test_remote_storage.py:826]: b5ab83a14

File: test_runner/regress/test_remote_storage.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply test_runner/regress/test_remote_storage.py]: b5ab83a14

File: pageserver/src/tenant/mgr.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs:1880]: This could had been done by polling on the metrics, but perhaps this is nicer, since there is pressure to implement this `wait_drop` anyways.

**Comment by koivunej** [comment]
> Failures on Postgres 14
> 
> - test_live_reconfig_get_evictions_low_residence_duration_metric_threshold: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-5108/6958243172/index.html#suites/4c54991a1bdef2d920f741ae5ea7d1a1/942ccacfd668dd7e)

From earlier test report. This one is really funky. The test creates extra tenant, but I don't understand why would compaction decide to download the only layer. Both tenants should have just one layer.

**Comment by koivunej** [comment]
> #### Failures on Postgres 15
> 
> * `test_timeline_delete_fail_before_local_delete`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-5108/6959328286/index.html#suites/f588e0a787c49e67b29490359c589fae/ef503a593e4652e6)

Fixed in [d501bd2](https://github.com/neondatabase/neon/pull/5108/commits/d501bd2734fe6268170ec2b67f2c1a5124c21c38)

File: pageserver/src/tenant/mgr.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs:1961]: [cbaf3bf](https://github.com/neondatabase/neon/pull/5108/commits/cbaf3bf4446ec5ec9a27cc0272996a7f27eebb4b)

File: test_runner/regress/test_remote_storage.py (hunk 4 above)
Comment (koivunej) [reply test_runner/regress/test_remote_storage.py]: This continues on a follow-up I had code ready: #5902.

**Comment by koivunej** [comment]
> #### Failures on Postgres 15
> 
> * `test_timeline_delete_fail_before_local_delete`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-5108/6959328286/index.html#suites/f588e0a787c49e67b29490359c589fae/ef503a593e4652e6)

This error makes no sense; the [d501bd2](https://github.com/neondatabase/neon/pull/5108/commits/d501bd2734fe6268170ec2b67f2c1a5124c21c38) most certainly fixes it after checking locally against the allure report log file, but the test execution is from a later commit. I don't understand.

**Comment by koivunej** [comment]
> #### Failures on Postgres 16
> 
> * `test_ondemand_download_timetravel[mock_s3]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-5108/6960529018/index.html#suites/b97efae3a617afb71cb8142f5afa5224/fc254918a12636cb)

Reported to #4112.

File: s3_scrubber/src/checks.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (problame) [reply s3_scrubber/src/checks.rs:146]: Ok.

---

Yet, I think the comment above this one (i.e., the one over the `gen < &index_part_generation`) already explains what orphans are.

This comment just rephrases / repeats most of it and adds confusion. Would rather see it deleted.

File: s3_scrubber/src/checks.rs [hunk 16]
```diff
@@ -136,24 +135,17 @@ pub(crate) async fn branch_cleanup_and_check_errors(
...
-                            "index_part.json does not contain layers from S3: {:?}",
-                            orphan_layers
-                                .iter()
-                                .map(|(layer_name, gen)| format!(
-                                    "{}{}",
-                                    layer_name.file_name(),
-                                    gen.get_suffix()
-                                ))
-                                .collect::<Vec<_>>(),
-                        ));
```
Comment (problame) [inline s3_scrubber/src/checks.rs:156]: Why did this need to go?

File: test_runner/fixtures/pageserver/http.py [hunk 17]
```diff
@@ -416,6 +425,11 @@ def timeline_delete(self, tenant_id: TenantId, timeline_id: TimelineId, **kwargs
     def timeline_gc(
         self, tenant_id: TenantId, timeline_id: TimelineId, gc_horizon: Optional[int]
     ) -> dict[str, Any]:
+        """
+        Unlike most handlers, this will wait for the layers to be actually
+        complete their shutdown up until registering them to the deletion
+        queue.
```
Comment (problame) [inline test_runner/fixtures/pageserver/http.py:431]: This is the first time I've heard of layers shutting down.

Please stick to established terminology, this just adds confusion.

File: test_runner/regress/test_remote_storage.py [hunk 18]
```diff
@@ -784,92 +782,87 @@ def test_compaction_waits_for_upload(
...
 
-        wait_for_last_flush_lsn(env, endpoint, tenant_id, timeline_id)
+    for name in upload_stuck_layers:
+        path = env.pageserver.timeline_dir(tenant_id, timeline_id) / name
+        assert path.exists(), "while uploads are stuck the layers should be present on disk"
 
-    checkpoint_result: queue.Queue[Optional[PageserverApiException]] = queue.Queue()
-    compact_result: queue.Queue[Optional[PageserverApiException]] = queue.Queue()
-    compact_barrier = threading.Barrier(2)
+    # now this will do the L0 => L1 compaction and want to remove our layers
```
Comment (problame) [inline test_runner/regress/test_remote_storage.py:820]:
```suggestion
    # now this will do the L0 => L1 compaction and want to remove our `upload_stuck_layers`
```

File: test_runner/regress/test_tenant_conf.py [hunk 19]
```diff
@@ -336,10 +336,15 @@ def test_live_reconfig_get_evictions_low_residence_duration_metric_threshold(
...
+        initial_tenant_conf={
+            # disable compaction so that it will not download the layer for repartitioning
+            "compaction_period": "0s"
+        }
+    )
     assert isinstance(env.pageserver_remote_storage, LocalFsStorage)
 
-    (tenant_id, timeline_id) = env.neon_cli.create_tenant()
+    (tenant_id, timeline_id) = env.initial_tenant, env.initial_timeline
     ps_http = env.pageserver.http_client()
```
Comment (problame) [inline test_runner/regress/test_tenant_conf.py:348]: This is a drive-by fix unrelated to the main change, right?

Would be nice to extract this into a preliminary PR, but, not insisting.


```suggestion
    env = neon_env_builder.init_start(
        initial_tenant_conf={
            # disable compaction so that it will not download the layer for repartitioning
            # (this test does assertions based on manual on-demand download & eviction)
            "compaction_period": "0s"
        }
    )
    assert isinstance(env.pageserver_remote_storage, LocalFsStorage)

    (tenant_id, timeline_id) = env.initial_tenant, env.initial_timeline
    ps_http = env.pageserver.http_client()
```

**Review by problame — APPROVED** [review]

**Comment by problame** [comment]
Before this PR, we had full mutual exlusivity between eviction and compaction.

Now, as I understand it, compaction's residency requirement still takes priority over eviction.

I guess that's an improvement, but really, we will need a mechanism for eviction to take priority over compaction as well, e.g., in disk-full situations. That, along with #5897 to prevent further data coming in, would actually be quite nice. 

Do we have an issue for such prioritization mechanism already?

File: pageserver/src/tenant/timeline.rs [hunk 20]
```diff
@@ -317,6 +309,24 @@ pub struct Timeline {
     /// Cancellation token scoped to this timeline: anything doing long-running work relating
     /// to the timeline should drop out when this token fires.
     pub(crate) cancel: CancellationToken,
+
+    /// Make sure we only have one running compaction at a time in tests.
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:313]: We don't have an API like do_gc without `cfg(feature = "testing")` or do you think we do have? I thought I checked.

File: s3_scrubber/src/checks.rs (hunk 16 above)
Comment (koivunej) [reply s3_scrubber/src/checks.rs:156]: Does not look familiar, need to check, I did do the merge main.

File: test_runner/regress/test_tenant_conf.py (hunk 19 above)
Comment (koivunej) [reply test_runner/regress/test_tenant_conf.py:348]: It was a test failure: https://github.com/neondatabase/neon/pull/5108#issuecomment-1823036202

Does your suggestion do something else than the comment? IMHO not stricly needed, it's quite clear what the test does.

Yeah, can split.

File: s3_scrubber/src/checks.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply s3_scrubber/src/checks.rs:146]: Unresolving, happy to remove extra comments.

File: pageserver/src/tenant/timeline.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:313]: Both of compaction driving endpoints are `testing_api_handler`.

File: pageserver/src/tenant/remote_timeline_client.rs [hunk 21]
```diff
@@ -841,6 +840,47 @@ impl RemoteTimelineClient {
...
+    /// operations but proper stop is yet to be called. On cancel the original or some later task
+    /// must call `stop` or `shutdown`.
+    pub(crate) async fn shutdown(self: &Arc<Self>) -> Result<(), StopError> {
+        let fut = {
+            let mut guard = self.upload_queue.lock().unwrap();
+            let upload_queue = match &mut *guard {
+                UploadQueue::Stopped(_) => return Ok(()),
+                UploadQueue::Uninitialized => return Err(StopError::QueueUninitialized),
+                UploadQueue::Initialized(ref mut init) => init,
+            };
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client.rs:855]: > except if we force shutdown the tokio runtime.

Even this is behind a `&mut` which is inaccessible with the lazy statics.

> I'd personally add a scopeguard or simliar to detect cancellation and panic in that case, but I have a hunch you wouldn't do that.

I had been thinking of a scopeguard to log, not to panic.

> So, let's just call it not cancellation safe, and also remove the comments in the function that "think through" this case.

I think it's fine to describe the non-trivial behaviour which would be caused along with the TODO about refactoring `launch_queued_tasks`.

What is the 'function that "think through" this case'?

File: pageserver/src/tenant/upload_queue.rs [hunk 22]
```diff
@@ -88,6 +88,14 @@ pub(crate) struct UploadQueueInitialized {
     /// bug causing leaks, then it's better to not leave this enabled for production builds.
     #[cfg(feature = "testing")]
     pub(crate) dangling_files: HashMap<LayerFileName, Generation>,
+
+    /// Set to true when we have inserted the first `UploadOp::Shutdown` into the `inprogress_tasks`.
+    pub(crate) shutting_down: bool,
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/upload_queue.rs:93]: Yes this must've been a left over from earlier deadlocky version, removing the first but keeping the non-mathascii version like:
```suggestion
    /// Set to true when we have inserted the `UploadOp::Shutdown` into the `inprogress_tasks`.
    pub(crate) shutting_down: bool,
```

File: pageserver/src/tenant/remote_timeline_client.rs [hunk 23]
```diff
@@ -1417,10 +1467,11 @@ impl RemoteTimelineClient {
     }
 
     /// Close the upload queue for new operations and cancel queued operations.
+    ///
     /// In-progress operations will still be running after this function returns.
     /// Use `task_mgr::shutdown_tasks(None, Some(self.tenant_id), Some(timeline_id))`
     /// to wait for them to complete, after calling this function.
-    pub fn stop(&self) -> Result<(), StopError> {
+    pub(crate) fn stop(&self) -> Result<(), StopError> {
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client.rs:1474]: c646dcf98d0c5cf246cb5d7a013d24e582ec9e1e

File: test_runner/regress/test_tenant_conf.py (hunk 19 above)
Comment (koivunej) [reply test_runner/regress/test_tenant_conf.py:348]: Split to #5926

**Comment by koivunej** [comment]
> Before this PR, we had full mutual exlusivity between eviction and compaction.
> 
> Now, as I understand it, compaction's residency requirement still takes priority over eviction.

I don't see how the "compaction needs L0s to be present" would had been changed in this PR.

> I guess that's an improvement, but really, we will need a mechanism for eviction to take priority over compaction as well, e.g., in disk-full situations. That, along with https://github.com/neondatabase/neon/issues/5897 to prevent further data coming in, would actually be quite nice.

the improvement is that there is no more mutual exclusion. agreed that there might be cases where one might wish for a single layer point of view that we get that layer forced away, however I think there is always a chance of evicting other layers (of possibly the same tenant).

> Do we have an issue for such prioritization mechanism already?

the Layer implementation works explicitly in the sense that evictions lose to downloads. no such issue exists.

File: test_runner/regress/test_remote_storage.py (hunk 18 above)
Comment (koivunej) [reply test_runner/regress/test_remote_storage.py:820]: Pushing soon 592c3244c, which is not this suggestion because there will be removal for the initdb layer as well.

File: test_runner/fixtures/pageserver/http.py (hunk 17 above)
Comment (koivunej) [reply test_runner/fixtures/pageserver/http.py:431]: d2aea5624 will talk just about deletion queue not shutdown.

File: s3_scrubber/src/checks.rs (hunk 16 above)
Comment (koivunej) [reply s3_scrubber/src/checks.rs:156]: https://github.com/neondatabase/neon/commit/d8c69e7c5dab097404d3800284044030bb4ed5a5 says "orphan layers are not errors". the idea is that we allow the orphan layers which are unlinked layers.

File: pageserver/src/tenant/upload_queue.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/upload_queue.rs]: bbc0f3081

File: s3_scrubber/src/checks.rs (hunk 16 above)
Comment (problame) [reply s3_scrubber/src/checks.rs]: Hm, I haven't followed scrubber too closely.
I.e., I don't know whether scrubber distinguishes between errors and "expected garbage".
Did @jcsp review this piece?

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/remote_timeline_client.rs:865]: discussed offline

File: s3_scrubber/src/checks.rs (hunk 16 above)
Comment (jcsp) [reply s3_scrubber/src/checks.rs]: It's okay to drop this -- the original code (not written by me) was basically asserting a totally quiescent state, where all layers were referenced by index and vice versa.  In real life, a layer can be un-referenced by the index during deletion, compaction, or initial creation.

Ideally we would warn on this case, or have a global flag for the  scrubber for how sensitive it should be (so that during certain tests we would run an "everything should be  perfect" mode).

Comment (problame) [reply s3_scrubber/src/checks.rs]: Yeah, that's how I remembered the plan for the scrubber: under normal operation, we shouldn't be leaking layers.
So, I guess we should have infra in the test framework to mark a test as "leaked layers are ok/expected".

@koivunej what tests are actually leaking layers with this PR?

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client.rs]: yeah I failed to understand the other function was a reference to "needing to make sure the other function does is not cancellable". this should be pushed now or will be in a sec.

File: s3_scrubber/src/checks.rs (hunk 16 above)
Comment (koivunej) [reply s3_scrubber/src/checks.rs]: Probably the ones which are using immediate stop or restart... Checking.

Comment (koivunej) [reply s3_scrubber/src/checks.rs]: Weird, no errors this time. I did not remove it for fun, it was causing failed tests, but it is possible these were additional errors to tests which are now benefitting from the timeline_gc waiting.

**Comment by koivunej** [comment]
> #### Failures on Postgres 16
> 
> * `test_emergency_relocate_with_branches_slow_replay[mock_s3]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-5108/7019598053/index.html#suites/9d2450a537238135fd4007859e09aca7/7506cf18c37c94a3)

Looks flaky but unsure why. Re-ran.

File: s3_scrubber/src/checks.rs (hunk 16 above)
Comment (koivunej) [reply s3_scrubber/src/checks.rs]: Left this out.
